services/worker-manager/src/providers/google.js

```diff
@@ -197,22 +198,59 @@ class GoogleProvider extends Provider {
     });
   }
 
+  /**
+   * Attempt to know how many "pending" instances total there
+   * are. There are a couple cases:
+   * 1. We have run a complete scanWorker loop since the last time we spun up
+   *    instances and therefore have an authoritative count on how many instances
+   *    are in a pending state. Just return that
+   * 2. We have not run a complete scanWorker loop since we last requested instances
+   *    so we have to return the sum of the authoritative pending instances from
+   *    the last loop and the number we think we've asked for since then.
+   * Note that since these two loops are not in-sync, this is not 100% accurate but
+   * this should do good enough of a job estimating for us to reliably have enough
+   * instances on hand but not over-commit.
+   */
+  _estimatePending({requested, pending}) {
+    if (pending.time >= requested.time) {
+      return pending.count;
+    }
+    return pending.count + requested.count;
+  }
+
   async provision({workerPool}) {
     const {workerPoolId} = workerPool;
 
-    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
-      await workerPool.modify(wt => {
-        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
-        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0;
-      });
+    let providerData = workerPool.providerData[this.providerId] || {};
+
+    // Migrate old workerpools to new config
+    if (typeof providerData.running === 'number') {
+      providerData.running = {count: providerData.running, time: new Date()};
     }
+
+    providerData.running = providerData.running || {count: 0, time: new Date()};
+    providerData.pending = providerData.pending || {count: 0, time: new Date()};
+    providerData.requested = providerData.requested || {count: 0, time: new Date()};
+
+    // TODO: Probably something to do with multiple configs and networks?
+    // TODO: And also multiple possible instance types :/
     const regions = workerPool.config.regions;
 
-    const toSpawn = await this.estimator.simple({
+    const toSpawn = Math.ceil((await this.estimator.simple({
       workerPoolId,
       ...workerPool.config,
-      running: workerPool.providerData[this.providerId].running,
-    });
+      runningCapacity: providerData.running.count,
+      pendingCapacity: this._estimatePending(providerData),
+    })) / workerPool.config.capacityPerInstance);
```

Comment:
maybe the division by `capacityPerInstance` and the `Math.ceil` operations should go into the estimator?